### ISSUE TITLE
Implement the Variational Quantum Linear Solver (VQLS)

### ISSUE THREAD

**Issue by nelimee** [Qiskit/qiskit-terra#5763]
The VQLS algorithm [Variational Quantum Linear Solver] aims at solving linear systems of equations using a variational approach.

I think adding VQLS to Qiskit would be beneficial as it would enrich the library with another variational algorithm. Moreover, linear systems of equations are a fundamental problem in a lot of areas and having such an algorithm implemented in Qiskit might attract some new users interested in this kind of problems (just like HHL).

## What is the expected behavior?

Being able to import and use the VQLS algorithm into Qiskit.

```python
from qiskit.aqua.algorithms.linear_solvers import VQLS
from qiskit.circuit.library import EfficientSU2
from qiskit.aqua.components.optimizers import COBYLA

from qiskit import BasicAer

qubit_number = 10

# Inspired from VQE signature found in 
# https://qiskit.org/documentation/stubs/qiskit.aqua.algorithms.VQE.html#qiskit.aqua.algorithms.VQE
vqls_instance = VQLS(
    decomposition={
        "circuits": get_decomposition(qubit_number),    # User-defined
        "coefficients": get_coefficients(qubit_number), # User-defined
    }, 
    var_form=EfficientSU2(qubit_number, reps=1), 
    optimizer=COBYLA()
)

result = vqls_instance.run(BasicAer.get_backend("statevector_simulator"))
print(result)
```

## Implementation steps

Implementing the VQLS algorithm in Qiskit boils down to implementing at least one of the two cost functions introduced by [Variational Quantum Linear Solver]:

- The global cost function
- The local cost function

Both cost functions requires to estimate "expectation values" of unitary matrices, i.e. <0| U |0> with U being a unitary matrix implemented as a quantum circuit.
This is done in [Variational Quantum Linear Solver] with the [Hadamard Test](https://en.wikipedia.org/wiki/Hadamard_test_(quantum_computation)). In addition, the authors of [Variational Quantum Linear Solver] also introduce the Hadamard-Overlap test, a specialised version of the Hadamard Test that solves a restricted problem (only works for some specific U, not for any U) by generating a shallower circuit, at the expense of using more qubits.

Implementing the Hadamard Test algorithm is required if we want to keep VQLS as generic as possible. Without an Hadamard Test implementation, the VQLS implementation would be limited to corner cases explained in [Methodology for replacing indirect measurements with direct measurements].
Implementing the Hadamard-Overlap test is optional, but allows to improve the shallowness of some of the generated quantum circuits, which is of particular interest for NISQ devices. 
Implementing the global cost function may not be a priority, as [Barren plateaus in quantum neural network training landscapes] showed that it will inevitably suffer from vanishing gradients.

So we are left with a minimal implementation:

- Hadamard Test
- VQLS local cost function

With possible improvements once 

- Hadamard-Overlap Test

is implemented.

## Side note

I wanted to start the discussion about implementing VQLS before implementing anything in Qiskit and wait for feedback. As such, I did not start to write any code for VQLS on Qiskit yet and i will not start until I have positive feedback from maintainers/regular contributors. 
But even if I did not start with Qiskit, I already have a working VQLS implementation written with another quantum computing framework. The Qiskit implementation might benefit from this experience.

[Variational Quantum Linear Solver]: https://arxiv.org/abs/1909.05820 "Variational Quantum Linear Solver"
[Methodology for replacing indirect measurements with direct measurements]: https://doi.org/10.1103/PhysRevResearch.1.013006 "Methodology for replacing indirect measurements with direct measurements"
[Barren plateaus in quantum neural network training landscapes]: https://www.nature.com/articles/s41467-018-07090-4 "Barren plateaus in quantum neural network training landscapes"

**Comment by woodsp-ibm** [comment]
For HHL rework see #5698

**Comment by stefan-woerner** [comment]
@nelimee Thanks for reaching out! Best wait for #5698 being merged. We are almost there so you can already look into it and let us know in case you have any questions.

**Comment by Mostafa-Atallah2020** [comment]
Hi, @nelimee are you still working on this ??

**Comment by nelimee** [comment]
I am not working on this anymore, and I do not have any code to share so you can consider that nothing has been done.

**Comment by Mostafa-Atallah2020** [comment]
Hi @stefan-woerner, @anedumla , @pbark I am asking to work on this issue.

**Comment by anedumla** [comment]
Hi @Mostafa-Atallah2020, recently we decided to deprecate the linear systems package from Qiskit for the time being. If somebody wants to implement VQLS it should also be separately from Qiskit, so I think we should close this issue.

**Comment by woodsp-ibm** [comment]
As Almudena mentioned the linear solvers package here was deprecated last release and will soon be removed. The code that was in Terra is now maintained elsewhere https://github.com/anedumla/quantum_linear_solvers As mentioned if you want to proceed I think its best to have this separate from Qiskit Terra at present. Now if you are interested I think its possible, if you create such an algorithm, to have it as part of the Qiskit Ecosystem, see https://qiskit.org/ecosystem/ Given the decision to deprecate and remove Iinear solvers I will close this issue.
